File: third/depend/MVS/Development/Samples/Python/ParametrizeCamera_LineScanIOSettings/ParametrizeCamera_LineScanIOSettings.py
# -- coding: utf-8 --

import sys
import msvcrt

sys.path.append("../MvImport")
from MvCameraControl_class import *


winfun_ctype = WINFUNCTYPE

stFrameInfo = POINTER(MV_FRAME_OUT_INFO_EX)
pData = POINTER(c_ubyte)
FrameInfoCallBack = winfun_ctype(None, pData, stFrameInfo, c_void_p)

def image_callback(pData, pFrameInfo, pUser):
        stFrameInfo = cast(pFrameInfo, POINTER(MV_FRAME_OUT_INFO_EX)).contents
        if stFrameInfo:
            print("get one frame: Width[%d], Height[%d], nFrameNum[%d]" % (stFrameInfo.nWidth, stFrameInfo.nHeight, stFrameInfo.nFrameNum))

CALL_BACK_FUN = FrameInfoCallBack(image_callback)

if __name__ == "__main__":

    try:
        # ch:初始化SDK | en: initialize SDK
        MvCamera.MV_CC_Initialize()

        deviceList = MV_CC_DEVICE_INFO_LIST()
        tlayerType = (MV_GIGE_DEVICE | MV_USB_DEVICE | MV_GENTL_CAMERALINK_DEVICE
                      | MV_GENTL_CXP_DEVICE | MV_GENTL_XOF_DEVICE)

        # ch:枚举设备 | en:Enum device
        ret = MvCamera.MV_CC_EnumDevices(tlayerType, deviceList)
        if ret != 0:
            print("enum devices fail! ret[0x%x]" % ret)
            sys.exit()

        if deviceList.nDeviceNum == 0:
            print("find no device!")
            sys.exit()

        print("Find %d devices!" % deviceList.nDeviceNum)

        for i in range(0, deviceList.nDeviceNum):
            mvcc_dev_info = cast(deviceList.pDeviceInfo[i], POINTER(MV_CC_DEVICE_INFO)).contents
            if mvcc_dev_info.nTLayerType == MV_GIGE_DEVICE or mvcc_dev_info.nTLayerType == MV_GENTL_GIGE_DEVICE:
                print("\ngige device: [%d]" % i)
                strModeName = ""
                for per in mvcc_dev_info.SpecialInfo.stGigEInfo.chModelName:
                    if per == 0:
                        break
                    strModeName = strModeName + chr(per)
                print("device model name: %s" % strModeName)

                nip1 = ((mvcc_dev_info.SpecialInfo.stGigEInfo.nCurrentIp & 0xff000000) >> 24)
                nip2 = ((mvcc_dev_info.SpecialInfo.stGigEInfo.nCurrentIp & 0x00ff0000) >> 16)
                nip3 = ((mvcc_dev_info.SpecialInfo.stGigEInfo.nCurrentIp & 0x0000ff00) >> 8)
                nip4 = (mvcc_dev_info.SpecialInfo.stGigEInfo.nCurrentIp & 0x000000ff)
                print("current ip: %d.%d.%d.%d\n" % (nip1, nip2, nip3, nip4))
            elif mvcc_dev_info.nTLayerType == MV_USB_DEVICE:
                print("\nu3v device: [%d]" % i)
                strModeName = ""
                for per in mvcc_dev_info.SpecialInfo.stUsb3VInfo.chModelName:
                    if per == 0:
                        break
                    strModeName = strModeName + chr(per)
                print("device model name: %s" % strModeName)

                strSerialNumber = ""
                for per in mvcc_dev_info.SpecialInfo.stUsb3VInfo.chSerialNumber:
                    if per == 0:
                        break
                    strSerialNumber = strSerialNumber + chr(per)
                print("user serial number: %s" % strSerialNumber)
            elif mvcc_dev_info.nTLayerType == MV_GENTL_CAMERALINK_DEVICE:
                print("\nCML device: [%d]" % i)
                strModeName = ""
                for per in mvcc_dev_info.SpecialInfo.stCMLInfo.chModelName:
                    if per == 0:
                        break
                    strModeName = strModeName + chr(per)
                print("device model name: %s" % strModeName)

                strSerialNumber = ""
                for per in mvcc_dev_info.SpecialInfo.stCMLInfo.chSerialNumber:
                    if per == 0:
                        break
                    strSerialNumber = strSerialNumber + chr(per)
                print("user serial number: %s" % strSerialNumber)
            elif mvcc_dev_info.nTLayerType == MV_GENTL_CXP_DEVICE:
                print("\nCXP device: [%d]" % i)
                strModeName = ""
                for per in mvcc_dev_info.SpecialInfo.stCXPInfo.chModelName:
                    if per == 0:
                        break
                    strModeName = strModeName + chr(per)
                print("device model name: %s" % strModeName)

                strSerialNumber = ""
                for per in mvcc_dev_info.SpecialInfo.stCXPInfo.chSerialNumber:
                    if per == 0:
                        break
                    strSerialNumber = strSerialNumber + chr(per)
                print("user serial number: %s" % strSerialNumber)
            elif mvcc_dev_info.nTLayerType == MV_GENTL_XOF_DEVICE:
                print("\nXoF device: [%d]" % i)
                strModeName = ""
                for per in mvcc_dev_info.SpecialInfo.stXoFInfo.chModelName:
                    if per == 0:
                        break
                    strModeName = strModeName + chr(per)
                print("device model name: %s" % strModeName)

                strSerialNumber = ""
                for per in mvcc_dev_info.SpecialInfo.stXoFInfo.chSerialNumber:
                    if per == 0:
                        break
                    strSerialNumber = strSerialNumber + chr(per)
                print("user serial number: %s" % strSerialNumber)

        nConnectionNum = input("please input the number of the device to connect:")

        if int(nConnectionNum) >= deviceList.nDeviceNum:
            print("input error!")
            sys.exit()

        nTriggerIndex = input("Please Input trigger selector index: 0-FrameBurstStart, 1-LineStart:")
        if int(nTriggerIndex) not in {0, 1}:
            print("input error!")
            sys.exit()

        # ch:创建相机实例 | en:Create Camera Object
        cam = MvCamera()

        # ch:选择设备并创建句柄 | en:Select device and create handle
        stDeviceList = cast(deviceList.pDeviceInfo[int(nConnectionNum)], POINTER(MV_CC_DEVICE_INFO)).contents

        ret = cam.MV_CC_CreateHandle(stDeviceList)
        if ret != 0:
            raise Exception("create handle fail! ret[0x%x]" % ret)

        # ch:打开设备 | en:Open device
        ret = cam.MV_CC_OpenDevice(MV_ACCESS_Exclusive, 0)
        if ret != 0:
            raise Exception("open device fail! ret[0x%x]" % ret)

        # ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
        if stDeviceList.nTLayerType == MV_GIGE_DEVICE or stDeviceList.nTLayerType == MV_GENTL_GIGE_DEVICE:
            nPacketSize = cam.MV_CC_GetOptimalPacketSize()
            if int(nPacketSize) > 0:
                ret = cam.MV_CC_SetIntValue("GevSCPSPacketSize", nPacketSize)
                if ret != 0:
                    print("Warning: Set Packet Size fail! ret[0x%x]" % ret)
            else:
                print("Warning: Get Packet Size fail! ret[0x%x]" % nPacketSize)

        # ch:线阵相机帧触发预设置 | en:Settings in FrameBurstStart mode
        if int(nTriggerIndex) == 0:
            # ch:设置触发选项为FrameBurstStart | en:Set trigger selector as FrameBurstStart
            ret = cam.MV_CC_SetEnumValue("TriggerSelector", 6)
            if ret != 0:
                raise Exception("set trigger selector fail! ret[0x%x]" % ret)

            # ch:设置触发模式为on | en:Set trigger mode as on
            ret = cam.MV_CC_SetEnumValue("TriggerMode", 1)
            if ret != 0:
                raise Exception("set trigger mode fail! ret[0x%x]" % ret)

            # ch:设置触发源为Line0 | en:Set trigger source as Line0
            ret = cam.MV_CC_SetEnumValue("TriggerSource", 0)
            if ret != 0:
                raise Exception("set trigger source fail! ret[0x%x]" % ret)
        # ch:线阵相机行触发预设置 | en:Settings in LineStart mode
        else:
            # ch:设置触发选项为LineStart | en:Set trigger selector as LineStart
            ret = cam.MV_CC_SetEnumValue("TriggerSelector", 9)
            if ret != 0:
                raise Exception("set trigger selector fail! ret[0x%x]" % ret)

            # ch:设置触发模式为on | en:Set trigger mode as on
            ret = cam.MV_CC_SetEnumValue("TriggerMode", 1)
            if ret != 0:
                raise Exception("set trigger mode fail! ret[0x%x]" % ret)

            # ch:设置触发源为EncoderModuleOut | en:Set trigger source as EncoderModuleOut
            ret = cam.MV_CC_SetEnumValue("TriggerSource", 6)
            if ret != 0:
                raise Exception("set trigger source fail! ret[0x%x]" % ret)

            # ch:设置编码器选项为Encoder0 | en:Set encoder selector as Encoder0
            ret = cam.MV_CC_SetEnumValue("EncoderSelector", 0)
            if ret != 0:
                raise Exception("set encoder selector fail! ret[0x%x]" % ret)

            # ch:设置编码器数据源A为Line1 | en:Set encoder source A as Line1
            ret = cam.MV_CC_SetEnumValue("EncoderSourceA", 1)
            if ret != 0:
                raise Exception("set encoder sourceA fail! ret[0x%x]" % ret)

            # ch:设置编码器数据源B为Line3 | en:Set encoder source B as Line3
            ret = cam.MV_CC_SetEnumValue("EncoderSourceB", 3)
            if ret != 0:
                raise Exception("set encoder sourceB fail! ret[0x%x]" % ret)

        # ch:注册抓图回调 | en:Register image callback
        ret = cam.MV_CC_RegisterImageCallBackEx(CALL_BACK_FUN, None)
        if ret != 0:
            raise Exception("register image callback fail! ret[0x%x]" % ret)

        # ch:开始取流 | en:Start grab image
        ret = cam.MV_CC_StartGrabbing()
        if ret != 0:
            raise Exception("start grabbing fail! ret[0x%x]" % ret)

        print("press a key to stop grabbing.")
        msvcrt.getch()

        # ch:停止取流 | en:Stop grab image
        ret = cam.MV_CC_StopGrabbing()
        if ret != 0:
            raise Exception("stop grabbing fail! ret[0x%x]" % ret)

        # ch:关闭设备 | Close device
        ret = cam.MV_CC_CloseDevice()
        if ret != 0:
            raise Exception("close device fail! ret[0x%x]" % ret)

        # ch:销毁句柄 | Destroy handle0
        cam.MV_CC_DestroyHandle()

    except Exception as e:
        print(e)
        cam.MV_CC_CloseDevice()
        cam.MV_CC_DestroyHandle()
    finally:
        # ch:反初始化SDK | en: finalize SDK
        MvCamera.MV_CC_Finalize()

File: third/depend/MVS/Development/Samples/Java/ParametrizeCamera_LineScanIOSettings/src/ParametrizeCamera_LineScanIOSettings.java
/***************************************************************************************************
 * @file      ParametrizeCamera_LineScanIOSettings.java
 * @breif     This sample shows how to modify FrameBurstStart and LineStart mode of the linear camera.
 * @author    *
 * @date      2024/07/18
 *
 * @warning
 * @version   V1.0.0  2024/07/18 create this file.
 * @since     2024/07/18
 **************************************************************************************************/

import java.io.*;
import java.util.ArrayList;
import java.util.Scanner;

import MvCameraControlWrapper.*;
import static MvCameraControlWrapper.MvCameraControlDefines.*;

public class ParametrizeCamera_LineScanIOSettings
{
	public static Scanner scanner;
    private static void printDeviceInfo(MV_CC_DEVICE_INFO stDeviceInfo)
    {
        if (null == stDeviceInfo) {
            System.out.println("stDeviceInfo is null");
            return;
        }

        if ((stDeviceInfo.transportLayerType == MV_GIGE_DEVICE)||(stDeviceInfo.transportLayerType == MV_GENTL_GIGE_DEVICE))
			{
            System.out.println("\tCurrentIp:       " + stDeviceInfo.gigEInfo.currentIp);
            System.out.println("\tModel:           " + stDeviceInfo.gigEInfo.modelName);
            System.out.println("\tUserDefinedName: " + stDeviceInfo.gigEInfo.userDefinedName);
        } else if (stDeviceInfo.transportLayerType == MV_USB_DEVICE) {
            System.out.println("\tUserDefinedName: " + stDeviceInfo.usb3VInfo.userDefinedName);
            System.out.println("\tSerial Number:   " + stDeviceInfo.usb3VInfo.serialNumber);
            System.out.println("\tDevice Number:   " + stDeviceInfo.usb3VInfo.deviceNumber);
        } 
		else if (stDeviceInfo.transportLayerType == MV_GENTL_CAMERALINK_DEVICE)
            {
                System.out.println("\tUserDefinedName: " + stDeviceInfo.cmlInfo.userDefinedName);
                System.out.println("\tSerial Number:   " + stDeviceInfo.cmlInfo.serialNumber);
                System.out.println("\tDevice Number:   " + stDeviceInfo.cmlInfo.DeviceID);
            }
            else if (stDeviceInfo.transportLayerType == MV_GENTL_CXP_DEVICE)
            {
                System.out.println("\tUserDefinedName: " + stDeviceInfo.cxpInfo.userDefinedName);
                System.out.println("\tSerial Number:   " + stDeviceInfo.cxpInfo.serialNumber);
                System.out.println("\tDevice Number:   " + stDeviceInfo.cxpInfo.DeviceID);
            }
            else if (stDeviceInfo.transportLayerType == MV_GENTL_XOF_DEVICE)
            {
                System.out.println("\tUserDefinedName: " + stDeviceInfo.xofInfo.userDefinedName);
                System.out.println("\tSerial Number:   " + stDeviceInfo.xofInfo.serialNumber);
                System.out.println("\tDevice Number:   " + stDeviceInfo.xofInfo.DeviceID);
            }
			else {
            System.err.print("Device is not supported! \n");
        }

        System.out.println("\tAccessible:      "
            + MvCameraControl.MV_CC_IsDeviceAccessible(stDeviceInfo, MV_ACCESS_Exclusive));
        System.out.println("");
    }

    private static void printFrameInfo(MV_FRAME_OUT_INFO stFrameInfo)
    {
        if (null == stFrameInfo)
        {
            System.err.println("stFrameInfo is null");
            return;
        }

        StringBuilder frameInfo = new StringBuilder("");
        frameInfo.append(("\tFrameNum[" + stFrameInfo.frameNum + "]"));
        frameInfo.append("\tWidth[" + stFrameInfo.width + "]");
        frameInfo.append("\tHeight[" + stFrameInfo.height + "]");
        frameInfo.append(String.format("\tPixelType[%#x]", stFrameInfo.pixelType.getnValue()));

        System.out.println(frameInfo.toString());
    }

    public static void saveDataToFile(byte[] dataToSave, int dataSize, String fileName)
    {
        OutputStream os = null;

        try
        {
			if((null == dataToSave)||(dataSize <= 0))
			{
				System.out.println("saveDataToFile param error.");
				return;
			}
            // create diractory
            File tempFile = new File("dat");
            if (!tempFile.exists()) 
            {
                tempFile.mkdirs();
            }

            os = new FileOutputStream(tempFile.getPath() + File.separator + fileName);
			if(null != os)
			{
				os.write(dataToSave, 0, dataSize);
                System.out.println("ConvertPixelType succeed.");
			}
            
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
        finally
        {
            // close file stream
            try 
            {
				if(os != null)
				{
					 os.close();
				}
               
            } catch (IOException e) 
            {
                e.printStackTrace();
            }
        }
    }

   public static int chooseCamera(ArrayList<MV_CC_DEVICE_INFO> stDeviceList)
    {
        if (null == stDeviceList)
        {
            return -1;
        }
        
        // Choose a device to operate
        int camIndex = -1;

        while (true)
        {
			System.out.print("Please input camera index:");
			if (scanner.hasNextInt()) 
			{
                try
                {
				   camIndex = scanner.nextInt();
                   if ((camIndex >= 0 && camIndex < stDeviceList.size()) || -1 == camIndex)
                   {
                       break;
                   }
                   else
                  {
                      System.out.println("Input error: " + camIndex + " Over Range:( 0 - " + (stDeviceList.size()-1) + " )");
                   }
                }
                catch (Exception e)
                {
			       System.out.println("Input not number.");
                   camIndex = -1;
                   break;
                }
			}
			else
			{
				camIndex = -1;
                break;
			}
			
        }
       if (-1 == camIndex) {
            System.out.println("Input error.exit");
            return camIndex;
        }

        if (0 <= camIndex && stDeviceList.size() > camIndex)
        {
            if ((MV_GIGE_DEVICE == stDeviceList.get(camIndex).transportLayerType)||(MV_GENTL_GIGE_DEVICE== stDeviceList.get(camIndex).transportLayerType))
            {
                System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).gigEInfo.userDefinedName);
            }
            else if (MV_USB_DEVICE == stDeviceList.get(camIndex).transportLayerType)
            {
                System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).usb3VInfo.userDefinedName);
            }
			 else if (stDeviceList.get(camIndex).transportLayerType == MV_GENTL_CAMERALINK_DEVICE)
            {
				System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).cmlInfo.DeviceID);
            }
            else if (stDeviceList.get(camIndex).transportLayerType == MV_GENTL_CXP_DEVICE)
            {
				System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).cxpInfo.DeviceID);
            }
            else if (stDeviceList.get(camIndex).transportLayerType == MV_GENTL_XOF_DEVICE)
            {
				System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).xofInfo.DeviceID);
            }
            else
            {
                System.out.println("Device is not supported.");
            }
        }
        else
        {
            System.out.println("Invalid index " + camIndex);
            camIndex = -1;
        }

        return camIndex;
    }

    public static void main(String[] args)
    {
        int nRet = MV_OK;
        int camIndex = -1;
        Handle hCamera = null;
		scanner = new Scanner(System.in);
        ArrayList<MV_CC_DEVICE_INFO> stDeviceList = null;

        do
        {
            System.out.println("SDK Version " + MvCameraControl.MV_CC_GetSDKVersion());

			// Initialize SDK
		    nRet = MvCameraControl.MV_CC_Initialize();
		    if (MV_OK != nRet)
		    {
			   System.err.printf("Initialize SDK fail! nRet [0x%x]\n\n",nRet);
               break;
		    }
			
            // Enumerate GigE and USB devices
            try
            {
                stDeviceList = MvCameraControl.MV_CC_EnumDevices(MV_GIGE_DEVICE | MV_USB_DEVICE | MV_GENTL_GIGE_DEVICE | MV_GENTL_CAMERALINK_DEVICE | MV_GENTL_CXP_DEVICE | MV_GENTL_XOF_DEVICE);
                if (0 >= stDeviceList.size())
                {
                    System.out.println("No devices found!");
                    break;
                }
                int i = 0;
                for (MV_CC_DEVICE_INFO stDeviceInfo : stDeviceList)
                {
                    System.out.println("[camera " + (i++) + "]");
                    printDeviceInfo(stDeviceInfo);
                }
            }
            catch (CameraControlException e)
            {
                System.err.println("Enumrate devices failed!" + e.toString());
                e.printStackTrace();
                break;
            }

            // choose camera
            camIndex = chooseCamera(stDeviceList);
            if (camIndex == -1)
            {
                break;
            }
            
            // Create device handle
            try
            {
                hCamera = MvCameraControl.MV_CC_CreateHandle(stDeviceList.get(camIndex));
            }
            catch (CameraControlException e)
            {
                System.err.println("Create handle failed!" + e.toString());
                e.printStackTrace();
                hCamera = null;
                break;
            }

            // Open selected device
            nRet = MvCameraControl.MV_CC_OpenDevice(hCamera);
            if (MV_OK != nRet)
            {
                System.err.printf("Connect to camera failed, errcode: [%#x]\n", nRet);
                break;
            }

			System.out.print("Please Input trigger selector index(0-1): 0-FrameBurstStart, 1-LineStart\n");
            int nTriggerSelector = -1;
			while(true)
			{
				if (scanner.hasNextInt()) 
			    {
					try
					{
						nTriggerSelector = scanner.nextInt();
						if ((nTriggerSelector >= 0 && nTriggerSelector < 2))
						{
							break;
						}
						else
						{
							System.out.println("Input error: " + nTriggerSelector + " Over Range:( 0 - " + "1" + " )");
							continue;
						}
					}
					catch (Exception e)
					{
						System.out.println("Input not number.");
						nTriggerSelector = -1;
						continue;
					}
				}
				else
				{
					nTriggerSelector = -1;
					continue;
				}
			}

            if (nTriggerSelector == 0)
		    {
			    // ch:设置触发选项为FrameBurstStart | en:Set trigger selector as FrameBurstStart
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "TriggerSelector", 6);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set Trigger Selector fail! nRet [0x%x]\n", nRet);
					break;
			    }

			    // ch:设置触发模式为on | en:Set trigger mode as on
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "TriggerMode", 1);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set Trigger Mode fail! nRet [0x%x]\n", nRet);
				    break;
			    }

			    // ch:设置触发源为Line0 | en:Set trigger source as Line0
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "TriggerSource", 0);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set Trigger source fail! nRet [0x%x]\n", nRet);
				    break;
			    }
		    }
		    else if (nTriggerSelector == 1)
		    {
			    // ch:设置触发选项为LineStart | en:Set trigger selector as LineStart
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "TriggerSelector", 9);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set Trigger Selector fail! nRet [0x%x]\n", nRet);
				    break;
			    }

			    // ch:设置触发模式为on | en:Set trigger mode as on
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "TriggerMode", 1);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set Trigger Mode fail! nRet [0x%x]\n", nRet);
				    break;
			    }

			    // ch:设置触发源为EncoderModuleOut | en:Set trigger source as EncoderModuleOut
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "TriggerSource", 6);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set Trigger source fail! nRet [0x%x]\n", nRet);
				    break;
			    }

			    // ch:设置编码器选项为Encoder0 | en:Set encoder selector as Encoder0
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "EncoderSelector", 0);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set encoder selector fail! nRet [0x%x]\n", nRet);
				    break;
			    }

			    // ch:设置编码器数据源A为Line1 | en:Set encoder source A as Line1
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "EncoderSourceA", 1);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set encoder sourceA fail! nRet [0x%x]\n", nRet);
				    break;
			    }

			    // ch:设置编码器数据源B为Line3 | en:Set encoder source B as Line3
			    nRet = MvCameraControl.MV_CC_SetEnumValue(hCamera, "EncoderSourceB", 3);
			    if (MV_OK != nRet)
			    {
                    System.err.printf("Set encoder sourceB fail! nRet [0x%x]\n", nRet);
				    break;
			    }
		    }
		    else
		    {
                System.err.printf("Input error!\n");
			    break;
		    }
		

            // Register image callback
            nRet = MvCameraControl.MV_CC_RegisterImageCallBack(hCamera, new CameraImageCallBack() {
                @Override
                public int OnImageCallBack(byte[] bytes,MV_FRAME_OUT_INFO mv_frame_out_info)
                {
                    printFrameInfo(mv_frame_out_info);
                    return 0;
                }
            });
            if (MV_OK != nRet)
            {
                System.err.printf("register image callback failed, errcode: [%#x]\n", nRet);
                break;
            }

            // Start grabbing images
            nRet = MvCameraControl.MV_CC_StartGrabbing(hCamera);
            if (MV_OK != nRet)
            {
                System.err.printf("Start Grabbing fail, errcode: [%#x]\n", nRet);
                break;
            }
			

			scanner.useDelimiter("");
            System.out.println("Press Enter to exit.");
            while(true)
            {
			    String input = scanner.nextLine();
                if(scanner.hasNextLine())
                {
                    break;
                }
			    else
                {
                    try {
                        Thread.sleep(1 * 10);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
					          	break;
               	 }
                }
                
            }
           
            // Stop grabbing
            nRet = MvCameraControl.MV_CC_StopGrabbing(hCamera);
            if (MV_OK != nRet)
            {
                System.err.printf("StopGrabbing failed, errcode: [%#x]\n", nRet);
                break;
            }
			
			// close device
            nRet = MvCameraControl.MV_CC_CloseDevice(hCamera);
            if (MV_OK != nRet)
            {
                System.err.printf("CloseDevice failed, errcode: [%#x]\n", nRet);
                break;
            }
			
			
        } while (false);

        if (null != hCamera)
        {
            // Destroy handle
            nRet = MvCameraControl.MV_CC_DestroyHandle(hCamera);
            if (MV_OK != nRet) {
                System.err.printf("DestroyHandle failed, errcode: [%#x]\n", nRet);
            }
        }
		MvCameraControl.MV_CC_Finalize();
		scanner.close();
    }
}
